graph/layout/eades.go

```diff
@@ -0,0 +1,171 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package layout
+
+import (
+	"math"
+
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/spatial/barneshut"
+	"gonum.org/v1/gonum/spatial/r2"
+)
+
+// EadesR2 implements the graph layout algorithm essentially as
+// described in "A heuristic for graph drawing", Congressus
+// numerantium 42:149-160.
+// The implementation here uses the Barnes-Hut approximation for
+// global repulsion calculation, and edge weights are considered
+// when calculating adjacent node attraction.
+type EadesR2 struct {
+	// Updates is the number of updates to perform.
+	Updates int
+
+	// Repulsion is the strength of the global
+	// repulsive force between nodes in the
+	// layout. It corresponds to C3 in the paper.
+	Repulsion float64
+
+	// Rate is the gradient descent rate. It
+	// corresponds to C4 in the paper.
+	Rate float64
+
+	// Theta is the Barnes-Hut theta constant.
+	Theta float64
+
+	// Src is the source of randomness used
+	// to initialize the nodes' locations. If
+	// Src is nil, the global random number
+	// generator is used.
+	Src rand.Source
+
+	nodes   graph.Nodes
+	indexOf map[int64]int
+
+	particles []barneshut.Particle2
+	forces    []r2.Vec
+}
+
+// Update is the EadesR2 spatial graph update function.
+func (u *EadesR2) Update(g graph.Graph, layout LayoutR2) bool {
+	if u.Updates <= 0 {
+		return false
+	}
+	u.Updates--
+
+	if !layout.IsInitialized() {
```

Comment:
Yes. It's pretty common for initial layouts to be one of random in a unit space (as is done here - gephi does something similar, but for reasons of java it doesn't run here at the moment though it did last week with no changes between then and now except maybe the jdk I have installed being updated), or layout out the nodes on a grid randomly with respect to connections. It would be possible to prearrange using Torgerson scaling, but that would make this particularly expensive for no real gain.

I think I'll just do the spreading more. Gephi uses 50x50 for random layouts, so maybe just start with that.

Reply:
I've just tried that and I don't think it's a good idea. Choosing a variety of other starting spreads gives bad layouts. This is not a justification for using this arbitrary unit random layout for this restricted set of input graphs, but rather a justification for not putting any real effort into making this layout work for all cases. The Eades layout here is really just to exercise the framework to see if it handles the kinds of things we need to do and worksish. Also note that with the current design it is possible to perform an initial random layout by writing an update function that assigns positions to nodes and then replacing the `Updater` value after that initial conditioning. I don't know if this is expecting too much of people.

I'm going to leave it as is for the moment.

Reply:
> I'm going to leave it as is for the moment.

Sure, I just wanted to finish commenting on this branch of the discussion.

> I don't know if this is expecting too much of people.

That's quite a hidden capability and not very elegant either, I'll think more about it when I've finished reading optimizer.go.

Reply:
>That's quite a hidden capability and not very elegant either, I'll think more about it when I've finished reading optimizer.go.

Mmmm. It has precedence in the standard library and it handles nicely a variety of layout problems that arise where an initial conditioning layout is needed and then a polishing layout is applied. I can't see a way of achieving that without a functionality like this.

Reply:
Why not initialize the layout before starting the Optimizer.Update loop? I'm afraid I'm looking at it from a wrong angle.